lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+//===- InferResets.cpp - Infer resets and add async reset -------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the InferResets pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/FieldRef.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "infer-resets"
+
+using llvm::BumpPtrAllocator;
+using llvm::MapVector;
+using llvm::SmallDenseSet;
+using llvm::SmallSetVector;
+using mlir::InferTypeOpInterface;
+using mlir::WalkOrder;
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+/// An absolute instance path.
+using InstancePathRef = ArrayRef<InstanceOp>;
+using InstancePath = SmallVector<InstanceOp>;
+
+template <typename T>
+static T &operator<<(T &os, InstancePathRef path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "/" << inst.name() << ":" << inst.moduleName();
+  return os;
+}
+
+static StringRef getTail(InstancePathRef path) {
+  if (path.empty())
+    return "$root";
+  auto last = path.back();
+  return last.name();
+}
+
+/// A reset domain.
+struct ResetDomain {
```

Comment:
Plz drop into an anonymous namespace, which will make all the stuff that uses this be implicitly static.